### PR TITLE
Hts 107 squashed

### PR DESCRIPTION
**Pull request by sureshhmrc**

### CODE REVIEW

File: app/uk/gov/hmrc/helptosavefrontend/config/frontendGlobal.scala [hunk 1]
```diff
@@ -14,42 +14,57 @@
...
-  extends DefaultFrontendGlobal {
+object FrontendGlobal extends DefaultFrontendGlobal with Redirects {
 
   override val auditConnector = FrontendAuditConnector
   override val loggingFilter = LoggingFilter
   override val frontendAuditFilter = AuditFilter
 
+  override def config: Configuration = play.api.Play.configuration
+
+  override lazy val env: Environment = Environment(Play.current.path, Play.current.classloader, Play.current.mode)
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/config/frontendGlobal.scala]: will fix these deprecated usages of `current` in a different PR

File: app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala [hunk 2]
```diff
@@ -18,60 +18,53 @@ package uk.gov.hmrc.helptosavefrontend.controllers
...
+class RegisterController @Inject()(override val messagesApi: MessagesApi,
+                                   htsService: HelpToSaveService,
+                                   nSAndIConnector: NSIConnector,
+                                   val sessionCacheConnector: SessionCacheConnector)
+  extends HelpToSaveAuth with I18nSupport {
+
+  def declaration: Action[AnyContent] = authorisedForHtsWithEnrolments(HtsAuthRule) {
+    implicit request ⇒
+      implicit userUrlWithNino ⇒
+        validateUser(userUrlWithNino.path.get, userUrlWithNino.nino.get).fold(
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala]: forgot to fix these `get` calls from an Option, will fix these

Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala]: Or you could change the type signature of `HtsActionWithEnrolments` from
```
case class UserDetailsUrlWithNino(path: Option[String], nino: Option[String])
HtsActionWithEnrolments = Request[AnyContent] => UserDetailsUrlWithNino => Future[Result]
```
to
```
case class UserDetailsUrlWithNino(path: String, nino: String)
HtsActionWithEnrolments = Request[AnyContent] => EitherT[Future,String,UserDetailsUrlWithNino] => Future[Result]
```
Then you could write
```
 def declaration: Action[AnyContent] = authorisedForHtsWithEnrolments(HtsAuthRule) {
    implicit request ⇒
      implicit userUrlWithNino ⇒
        for{
         UserDetailsUrlWithNino(path, nino) <- userUrlWithNino
         result                             <- validateUser(path, nino)
        } yield result.fold(..., ...)
```

File: app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala [hunk 3]
```diff
@@ -125,26 +123,9 @@ class RegisterController @Inject()(val sessionCacheConnector: SessionCacheConnec
     * been performed and the eligibility check is positive. This returns [[None]]
     * if all the above has successfully been performed and the eligibility check is negative.
     */
-  private def validateUser(authContext: AuthContext)(implicit hc: HeaderCarrier): Result[Option[UserInfo]] = for {
-    nino ← EitherT.fromOption[Future](retrieveNino(authContext), "Unable to retrieve NINO")
-    userInfo ← userInfoService.getUserInfo(authContext, nino)
-    eligible ← eligibilityConnector.checkEligibility(nino)
+  private def validateUser(userDetailsUri: String, nino: String)(implicit hc: HeaderCarrier): Result[Option[UserInfo]] = for {
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala]: I know I wrote this, but `validateUser` can probably be called something better like `performEligibilityCheck` or something

File: app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala (hunk 2 above)
Comment (cavster) [inline app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala]: Awesome

File: app/uk/gov/hmrc/helptosavefrontend/controllers/HelpToSaveAuth.scala [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+            val userUrlWithNino = (userDetailsUrl, nino) match {
+              case (Some(url), Some(ni)) ⇒ Some(UserDetailsUrlWithNino(url, ni))
+              case _ ⇒ None
+            }
+
+            action(request)(userUrlWithNino)
+
+        }.recover {
+        case _: NoActiveSession ⇒ toGGLogin(HtsDeclarationUrl)
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavefrontend/controllers/HelpToSaveAuth.scala]: this recover block is basically duplicated in `authorisedForHts` - create a separate method maybe?

File: app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala [hunk 5]
```diff
@@ -18,60 +18,58 @@ package uk.gov.hmrc.helptosavefrontend.controllers
...
+                                   val sessionCacheConnector: SessionCacheConnector,
+                                   implicit val app: Application)
+  extends HelpToSaveAuth(app) with I18nSupport {
+
+  def declaration: Action[AnyContent] = authorisedForHtsWithEnrolments {
+    implicit request ⇒
+      implicit userUrlWithNino ⇒
+        val userInfo = for {
+          userUrlWithNino <- EitherT.fromOption[Future](userUrlWithNino, "could not retrieve either userDetailsUrl or NINO from auth")
+          result <- checkEligibility(userUrlWithNino.path, userUrlWithNino.nino)
```
Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala]: Do we want to put the session cache code in this for comprehension? e.g.
```
for {
...
result <- checkEligibility(userUrlWithNino.path, userUrlWithNino.nino)
_       <- f(result)
} yield result

...
def f(userDetails: Option[UserInfo]): EitherT[Future,String,CacheMap] = {
  EitherT.fromOption[Future](userDetails, "No user details to write to key-store")
    .flatMapF(details ⇒
      sessionCacheConnector.put(HTSSession(Some(details)))
        .map[Either[String,CacheMap]](Right(_))
        .recover{ case e ⇒ Left(s"Could not write to key-store: ${e.getMessage}")}
    )
}
```
This way we make sure we have stored the information in key-store before proceeding

Comment (AndyHWChung) [inline app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala]: Also, I don't think `HTSSession` should have an `Option[UserInfo]` in it, it should have just a `UserInfo` in it - why would we want to write `None` to keystore? But this isn't part of this PR

Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/controllers/RegisterController.scala]: @AndyHWChung  sure, I will fix the above in the coming PR and merge the current one, is that ok ?

**Comment by AndyHWChung** [comment]
LGTM

**Pull request merged**